lockin/src/lib.rs

```diff
@@ -0,0 +1,973 @@
+#![no_std]
+use core::cmp::Ord;
+use core::convert::From;
+use core::ops::SubAssign;
+use core::f32::consts::PI;
+
+extern crate libm;
+
+mod trig;
+use trig::{atan2, cos, sin};
+
+// TODO temporary, pending new DMA code
+pub const SAMPLE_BUFFER_SIZE: usize = 16;
+pub const TSTAMP_BUFFER_SIZE: usize = SAMPLE_BUFFER_SIZE / 2;
+pub const OUTPUT_BUFFER_SIZE: usize = 1;
+
+/// Slow external reference edge timestamps.
+#[derive(Copy, Clone)]
+pub struct TimeStamp {
+    // Timestamp value.
+    pub count: u32,
+    // Number of sequences before the current one that the timestamp
+    // occurred. A sequence is a set of `SAMPLE_BUFFER_SIZE` ADC
+    // samples. E.g., a timestamp from the current sequence has this
+    // set to 0, a timestamp from the previous sequence has this set
+    // to 1, etc. A value of -1 indicates an invalid timestamp (i.e.,
+    // one that has not yet been set).
+    pub sequences_old: i32,
+}
+
+impl TimeStamp {
+    // Initialize a new TimeStamp.
+    pub fn new() -> Self {
+        TimeStamp {
+            count: 0,
+            sequences_old: -1,
+        }
+    }
+
+    /// Increments `sequences_old` if TimeStamp is valid. This is
+    /// called at the end of processing a sequence of ADC samples.
+    pub fn new_sequence(&mut self) {
+        if self.sequences_old != -1 {
+            self.sequences_old += 1;
+        }
+    }
+
+    /// Returns true if TimeStamp is valid.
+    pub fn is_valid(&self) -> bool {
+        self.sequences_old != -1
+    }
+
+    /// Set a new count value for TimeStamp. This also changes
+    /// `sequences_old` to 0 to indicate the TimeStamp belongs to the
+    /// current processing sequence.
+    ///
+    /// # Arguments
+    ///
+    /// * `newval` - New count value.
+    pub fn new_count(&mut self, newval: u16) {
+        self.count = newval as u32;
+        self.sequences_old = 0;
+    }
+}
+
+/// Compute the reference signal phase increment for each ADC sample
+/// and use them to demodulate the ADC inputs.
+macro_rules! prefilt_no_decimate {
+    ( $adc_samples:expr, $ref_tstamps:expr, $valid_tstamps:expr, $phi:expr, $tadc:expr, $fscale:expr, $tstamps_mem:expr) => {{
```

Comment:
This is a macro because it can return prematurely when `ts_valid_count < 2`. That return needs to be from the function in which the macro is invoked. Is there a better way to handle this?

Reply:
I would highly recommend insted using a `Result()` return value. The function that calls this can then check the return value for an `Ok(data)` or an `Err(Error::NotReady)`. It cleans up the code a lot here.

Reply:
I've done a major structural rewrite of this code (in line with your suggestions) and so this (and a bunch of other parts of this code) have been changed. I'm still cleaning it up, but hopefully some commits will be ready to push soon.